cmd/thanos/receive.go

```diff
@@ -196,18 +202,54 @@ func runReceive(
 		return errors.Wrap(err, "parse relabel configuration")
 	}
 
-	dbs := receive.NewMultiTSDB(
-		conf.dataDir,
-		logger,
-		reg,
-		tsdbOpts,
-		lset,
-		conf.tenantLabelName,
-		bkt,
-		conf.allowOutOfOrderUpload,
-		hashFunc,
+	// Moving these variable up, as there are few varibales which depend on these below it.
+	grpcProbe := prober.NewGRPC()
+	httpProbe := prober.NewHTTP()
+	statusProber := prober.Combine(
+		httpProbe,
+		grpcProbe,
+		prober.NewInstrumentation(comp, logger, extprom.WrapRegistererWithPrefix("thanos_", reg)),
+	)
+
+	// Since we need to initialize writer variable before-hand in order to prevent the error in enable ingestion block
+	var (
+		// hashringChangedChan signals when TSDB needs to be flushed and updated due to hashring config change.
+		hashringChangedChan = make(chan struct{}, 1)
+
+		dbs    *receive.MultiTSDB
+		writer *receive.Writer
 	)
-	writer := receive.NewWriter(log.With(logger, "component", "receive-writer"), dbs, conf.writerInterning)
+	// This block is to upload from TSDB where as, previous was to directly write to object storage from prometheus
+
+	if enableIngestion {
+		// Only if ingestion is enabled, tsdb should be initialized and the upload to tsdb should take place. ingestion or ingestion+router.
+		// Initializing database variable.
+		dbs = receive.NewMultiTSDB(
+			conf.dataDir,
+			logger,
+			reg,
+			tsdbOpts,
+			lset,
+			conf.tenantLabelName,
+			bkt,
+			conf.allowOutOfOrderUpload,
+			hashFunc,
+		)
+		// write to TSDB?
```

Comment:
@Kartik-Garg Who is this question for?

Reply:
@samkulkarni20 I have just left comments for my understanding